### PR TITLE
Feature/experiment manager

### PR DESCRIPTION
**Pull request by lorenzomammana**
## Summary

New features added with this draft PR:
- Add hash column to datamodule data field, save data field as csv during prepare_data
- Integrate mlflow autolog to track lightning checkpoints automatically
- Integrate mlflow logging of torchscript models at the end of experiment if upload_artifacts is enabled.

## Type of Change

Please select the one relevant option below:

- New feature (non-breaking change that adds functionality)

## Checklist

Please confirm that the following tasks have been completed:

- [ ] I have tested my changes locally and they work as expected. (Please describe the tests you performed.)
- [ ] I have added unit tests for my changes, or updated existing tests if necessary.
- [ ] I have updated the documentation, if applicable.
- [X] I have installed pre-commit and run locally for my code changes.

I'm using this PR as an experiment to go towards smaller pr with partially implemented tasks so we can try to review it faster and discuss some points.

For example I don't like how autolog is ignoring the MlflowLogger log_model = False!

That's why is in draft mode

### CODE REVIEW

**Comment by lorenzomammana** [comment]
I need to add documentation, but first I would like to review the implementation for autolog

File: docs/tutorials/tasks.md [hunk 1]
```diff
@@ -48,4 +48,4 @@ graph LR;
 Most of these tasks have an associated evaluation task used for inference.
 ## Adding new tasks
 
-If your require the `pytorch-lightning` library, you can add a new task by extending the [`LightningTask`][quadra.base.LightningTask] class. Otherwise, you can simply start implementing a new task by extending the `Task` class.
```
Comment (AlessandroPolidori) [inline docs/tutorials/tasks.md:51]: If you* require

File: quadra/datamodules/base.py [hunk 2]
```diff
@@ -46,6 +51,49 @@ def __new__(cls, name, bases, dct):
         return super().__new__(cls, name, bases, dct)
 
 
+def compute_file_hash(path: str, hash_size: int) -> str:
```
Comment (rcmalli) [inline quadra/datamodules/base.py:54]: If we cannot put any number for hash size, typing will be confusing. What about using [Literal Typings](https://peps.python.org/pep-0586/) ?

Comment (lorenzomammana) [reply quadra/datamodules/base.py:54]: I've already used Literal in the datamodule typings actually 😅 Makes sense

File: quadra/utils/export.py [hunk 3]
```diff
@@ -1,14 +1,19 @@
...
 
 import torch
 from anomalib.models.cflow import CflowLightning
 from torch import nn
 from torch.jit._script import RecursiveScriptModule
 
-from quadra.utils.utils import get_logger
 
-log = get_logger(__name__)
+def safe_get_logger() -> Logger:
```
Comment (rcmalli) [inline quadra/utils/export.py:11]: Okay, We need definitely work on loggers

File: quadra/utils/mlflow.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+import torch
+from torch import nn
+
+NnModuleT = TypeVar("NnModuleT", bound=nn.Module)
+
+
+@torch.inference_mode()
+def infer_signature_torch(model: NnModuleT, data: List[torch.Tensor]) -> Optional[ModelSignature]:
+    """Infer signature for a PyTorch/Torchscript model."""
+    from aigo_research.utils.utils import get_logger  # pylint: disable=[import-outside-toplevel]
```
Comment (rcmalli) [inline quadra/utils/mlflow.py:20]: 💯

File: quadra/utils/model_manager.py [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+            raise ImportError("Mlflow is not available, please install it with pip install mlflow")
+
+        if os.getenv("MLFLOW_TRACKING_URI") is None:
+            raise ValueError("MLFLOW_TRACKING_URI environment variable is not set")
+
+        mlflow.set_tracking_uri(os.getenv("MLFLOW_TRACKING_URI"))
+        self.client = MlflowClient()
+
+    def register_model(
+        self, model_location: str, model_name: str, tags: Optional[Dict[str, Any]] = None, description: str = ""
```
Comment (rcmalli) [inline quadra/utils/model_manager.py:74]: Maybe the description should be mandatory instead of having the default value of the empty string. If we want it to be optional, we can set it to be `None`

**Comment by rcmalli** [comment]
For me, it looks super cool overall! I will also run some experiments with it.

File: quadra/utils/model_manager.py (hunk 5 above)
Comment (lorenzomammana) [reply quadra/utils/model_manager.py:74]: I'll enforce it to be mandatory even tough I'm not a huge fan, but it probably makes more senso to always require a description

File: quadra/utils/model_manager.py [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+        mode: str = "max",
+        model_path: str = "model",
+    ) -> Any:
+        """Register the best model from an experiment."""
+
+
+class MlflowModelManager(AbstractModelManager):
+    """Model manager for Mlflow."""
+
+    def __init__(self):
```
Comment (rcmalli) [inline quadra/utils/model_manager.py:65]: It could be nice to let users pass Tracking URI as an argument.

**Comment by rcmalli** [comment]
If it is okay, we can align all API for description argument and make it not optional. Do you think there will be a problem? So far it is not optional only  for `register_model`.

File: quadra/utils/model_manager.py [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+        if not MLFLOW_AVAILABLE:
+            raise ImportError("Mlflow is not available, please install it with pip install mlflow")
+
+        if os.getenv("MLFLOW_TRACKING_URI") is None:
+            raise ValueError("MLFLOW_TRACKING_URI environment variable is not set")
+
+        mlflow.set_tracking_uri(os.getenv("MLFLOW_TRACKING_URI"))
```
[the remark this answers is not in the document]
Comment (lorenzomammana) [reply quadra/utils/model_manager.py:72]: I don't get the first parte, there's a check before to see if it is not None, what am I missing

**Comment by lorenzomammana** [comment]
Two checks are missing, we need to provide a way to enable or disable the upload of the export on mlflow (we may take advantage of the refactoring to the export parameter in input shape branch) and understand what's wrong with patches as I think tests are failing

**Review by rcmalli — APPROVED** [review]
